### PR TITLE
only handleErr on last retry failure

### PR DESCRIPTION
**Pull request by bparees**

### CODE REVIEW

**Comment by bparees** [comment]
@smarterclayton ptal.  I did not implement whitelisting of events because i don't know what that list would be currently.

File: pkg/build/controller/factory/factory.go [hunk 1]
```diff
@@ -158,8 +159,15 @@ func (factory *BuildPodControllerFactory) Create() controller.RunnableController
...
 			cache.MetaNamespaceKeyFunc,
 			func(obj interface{}, err error, retries controller.Retry) bool {
-				kutil.HandleError(err)
-				return retries.Count < maxRetries
+				pod := obj.(*kapi.Pod)
+				if retries.Count < maxRetries {
+					glog.V(3).Infof("Retrying BuildPod update event %s/%s: %v", pod.Namespace, pod.Name, err)
+					return true
+				} else {
+					glog.V(3).Infof("Giving up retrying BuildPod update event %s/%s: %v", pod.Namespace, pod.Name, err)
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: No else when you return in the if

File: pkg/build/controller/factory/factory.go [hunk 2]
```diff
@@ -230,11 +238,20 @@ func (factory *ImageChangeControllerFactory) Create() controller.RunnableControl
...
 				if _, isFatal := err.(buildcontroller.ImageChangeControllerFatalError); isFatal {
+					glog.V(3).Infof("Will not retry fatal error for ImageStream update event %s/%s: %v", imageStream.Namespace, imageStream.Name, err)
+					kutil.HandleError(err)
+					return false
+				}
+				if retries.Count < maxRetries {
+					glog.V(4).Infof("Retrying ImageStream update event %s/%s: %v", imageStream.Namespace, imageStream.Name, err)
+					return true
+				} else {
+					glog.V(3).Infof("Giving up retrying ImageStream update event %s/%s: %v", imageStream.Namespace, imageStream.Name, err)
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: Same

**Comment by bparees** [comment]
updated.

File: pkg/build/controller/factory/factory.go [hunk 3]
```diff
@@ -230,11 +238,20 @@ func (factory *ImageChangeControllerFactory) Create() controller.RunnableControl
...
 			cache.MetaNamespaceKeyFunc,
 			func(obj interface{}, err error, retries controller.Retry) bool {
-				kutil.HandleError(err)
+				imageStream := obj.(*imageapi.ImageStream)
 				if _, isFatal := err.(buildcontroller.ImageChangeControllerFatalError); isFatal {
+					glog.V(3).Infof("Will not retry fatal error for ImageStream update event %s/%s: %v", imageStream.Namespace, imageStream.Name, err)
+					kutil.HandleError(err)
+					return false
+				}
+				if retries.Count < maxRetries {
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: Invert this condition - make the handleerror the block, and return true as the end of the function

**Comment by bparees** [comment]
updated.

**Comment by bparees** [comment]
[test]

**Comment by smarterclayton** [comment]
So fresh and so clean [merge]

**Comment by bparees** [comment]
flake?  [merge]

**Comment by bparees** [comment]
[merge]

**Comment by bparees** [comment]
[INFO] Verifying the docker-registry is up at 172.30.189.33
ERROR: gave up waiting for http://172.30.189.33:5000/healthz

[FAIL] !!!!! Test Failed !!!!

[merge]

**Comment by bparees** [comment]
\o/

**Pull request merged**